### PR TITLE
[WIP] EN-261 Capsule vs Capsule collision

### PR DESCRIPTION
**Pull request by vladimir-tikhonov**
* Capsule vs Capsule collision
* Capsule vs Convex Hull collision
* Physics stability improvements

### CODE REVIEW

File: src/classes/Collision/CapsuleConvexHull.js [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        }
+
+        innerSegmentStart.set( 0, capsule.shape.half_height, 0 );
+        innerSegmentEnd.set( 0, -capsule.shape.half_height, 0 );
+        capsule.transform.transformVector3( innerSegmentStart );
+        capsule.transform.transformVector3( innerSegmentEnd );
+
+        // If closestPointOnHull is not null then inner segment does not collide with convex hull.
+        // This can mean two things - either there is no collision at all or there is a shallow collision.
+        var closestPointOnHull = Goblin.GjkEpa.findClosestPointOnObjectToLineSegment( convexHull, innerSegmentStart, innerSegmentEnd );
```
Comment (yard) [inline src/classes/Collision/CapsuleConvexHull.js:29]: Does this method use GJK/EPA at all? Or it simply sits there?

File: src/classes/Collision/CapsuleConvexHull.js [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    var closestPointOnInnerSegment = new Goblin.Vector3();
+    var contactNormal = new Goblin.Vector3();
+    var invertedRotationQuaternion = new Goblin.Quaternion();
+    var referencePlaneNormal = new Goblin.Vector3();
+    var clippedInnerSegmentStart = new Goblin.Vector3();
+    var clippedInnerSegmentEnd = new Goblin.Vector3();
+    var innerSegmentDirection = new Goblin.Vector3();
+    var epsilon = 0.01;
+
+    return function( capsule, convexHull, closestPointOnHull, innerSegmentStart, innerSegmentEnd, doLightweightCollision ) {
```
Comment (yard) [inline src/classes/Collision/CapsuleConvexHull.js:49]: Comments please

File: src/classes/Collision/CapsuleConvexHull.js [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contactAtTheStart, clippedInnerSegmentStart, contactAtTheStart.contact_point_in_b );
+        Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contactAtTheEnd, clippedInnerSegmentEnd, contactAtTheEnd.contact_point_in_b );
+
+        return [ contactAtTheStart, contactAtTheEnd ];
+    };
+} )();
+
+Goblin.Collision.capsuleConvexHull._deepCapsuleConvexHull = ( function() {
+    return function( capsule, convexHull, doLightweightCollision ) {
+        /**
```
Comment (yard) [inline src/classes/Collision/CapsuleConvexHull.js:142]: Are you absolutely sure we need these comments? I mean it's kinda obvious which object is returned, and, in worst case, we could use a normal // comment instead to note take 3 lines for a type hint.

File: src/classes/Collision/CapsuleConvexHull.js [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        if ( Math.abs( dotBetweenInnerSegmentDirectionAndContactNormal ) > epsilon ) {
+            Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contact, closestPointOnInnerSegment, closestPointOnHull );
+            return [ contact ];
+        }
+
+        invertedRotationQuaternion.invertQuaternion( convexHull.rotation );
+        invertedRotationQuaternion.transformVector3Into( contactNormal, referencePlaneNormal );
+        referencePlaneNormal.scale( -1 );
+        var referenceFace = null;
+        for ( var i = 0; i < convexHull.shape.faces.length; i++ ) {
```
Comment (yard) [inline src/classes/Collision/CapsuleConvexHull.js:87]: I bet this part will repeat over and over in many places - shouldn't we extact it to a convex hull method?

File: src/classes/Collision/CapsuleConvexHull.js [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            contact.is_lightweight = true;
+            return [ contact ];
+        }
+
+        return null;
+    };
+} )();
+
+/**
+ *
```
Comment (yard) [inline src/classes/Collision/CapsuleConvexHull.js:341]: Please specify what does it do :)

File: src/classes/Collision/CapsuleConvexHull.js [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+    contact.contact_point.addVectors( contact.contact_point_in_a, contact.contact_point_in_b );
+    contact.contact_point.scale( 0.5 );
+
+    contact.object_a.transform_inverse.transformVector3( contact.contact_point_in_a );
+    contact.object_b.transform_inverse.transformVector3( contact.contact_point_in_b );
+};
+
+/**
+ *
```
Comment (yard) [inline src/classes/Collision/CapsuleConvexHull.js:361]: And this one as well - and any other without a description :)

**Review by yard — CHANGES_REQUESTED** [review]

File: src/classes/Collision/CapsuleConvexHull.js (hunk 6 above)
Comment (vladimir-tikhonov) [reply src/classes/Collision/CapsuleConvexHull.js:361]: It's a little bit early to review this pull request - that's why it's marked as WIP and don't have assigned reviewers 😄   Nevertheless, I will address all your comments

File: src/classes/Collision/CapsuleConvexHull.js (hunk 1 above)
Comment (vladimir-tikhonov) [reply src/classes/Collision/CapsuleConvexHull.js:29]: Yes, one of the main purposes of GJK (which we didn't use before) - is to find closest points between objects

File: src/classes/Collision/CapsuleConvexHull.js (hunk 3 above)
Comment (vladimir-tikhonov) [reply src/classes/Collision/CapsuleConvexHull.js:142]: I just like a little bit of the strong typings / type hints. However, I 100% agree that such comments take up a lot of space. Will remove

File: src/classes/Collision/CapsuleConvexHull.js (hunk 4 above)
Comment (vladimir-tikhonov) [reply src/classes/Collision/CapsuleConvexHull.js:87]: Not sure what part are you referring to, but all the code in ±5 lines from there is completely unique

File: src/classes/Collision/CapsuleConvexHull.js (hunk 5 above)
Comment (vladimir-tikhonov) [reply src/classes/Collision/CapsuleConvexHull.js:341]: fixed

File: src/classes/Collision/CapsuleConvexHull.js (hunk 6 above)
Comment (vladimir-tikhonov) [reply src/classes/Collision/CapsuleConvexHull.js:361]: fixed

File: src/classes/Collision/CapsuleConvexHull.js (hunk 2 above)
Comment (vladimir-tikhonov) [reply src/classes/Collision/CapsuleConvexHull.js:49]: done

**Pull request closed without merging**